pkg/operation/botanist/secrets.go

```diff
@@ -760,27 +761,33 @@ func (b *Botanist) fetchExistingSecrets() (map[string]*corev1.Secret, error) {
 	return existingSecretsMap, nil
 }
 
-// Previously, we have used the same certificate authority in all places. Now, we are using dedicated CAs for the
-// different components. Gardener does not re-create/generate certificates/secrets if they already exist. Hence,
-// we have to delete the existing certificates to let them re-created and re-signed by a new CA.
-// This can be removed in a future version. See details: https://github.com/gardener/gardener/pull/353
-func (b *Botanist) deleteOldCertificates(existingSecretsMap map[string]*corev1.Secret) error {
-	var dedicatedCASecrets = map[string][]string{
-		caETCD:       []string{"etcd-server-tls", "etcd-client-tls"},
-		caFrontProxy: []string{"kube-aggregator"},
-		caKubelet:    []string{"kube-apiserver-kubelet"},
-	}
-	for caName, secrets := range dedicatedCASecrets {
-		for _, secretName := range secrets {
-			if _, ok := existingSecretsMap[caName]; !ok {
-				if err := b.K8sSeedClient.DeleteSecret(b.Shoot.SeedNamespace, secretName); err != nil && !apierrors.IsNotFound(err) {
-					return err
-				}
-				delete(existingSecretsMap, secretName)
-			}
-		}
+// Delete the etcd server certificate if it has been generated by an old version
+// of Gardener (not enough SANS).
+func (b *Botanist) deleteOldETCDServerCertificate(existingSecretsMap map[string]*corev1.Secret) error {
+	secret, ok := existingSecretsMap[certificateETCDServer]
+	if !ok {
+		return nil
 	}
 
+	val, err := secrets.LoadCertificate(certificateETCDServer, secret.Data[secrets.DataKeyPrivateKey], secret.Data[secrets.DataKeyCertificate])
+	if err != nil {
+		return err
+	}
+	certificate, ok := val.(*secrets.Certificate)
+	if !ok {
+		return fmt.Errorf("could not convert certificate into *secrets.Certificate struct")
+	}
+
+	if crt := certificate.Certificate; crt != nil && reflect.DeepEqual(crt.DNSNames, dnsNamesForEtcd(b.Shoot.SeedNamespace)) {
```

Comment:
isn't it better to use `sets`  and `HasAll` from apimachinery for this?
`reflect.DeepEqual` will fail when comparing `["a", "b"]` and `["b", "a"]` (slice order)

Reply:
Sure, agreed. Please check again.